rust/hyperlane-base/src/contract_sync/cursors/mod.rs

```diff
@@ -242,27 +261,36 @@ where
         } else {
             Duration::from_secs(0)
         };
+        Ok(eta)
+    }
+}
+
+#[async_trait]
+impl<T> ContractSyncCursor<T> for RateLimitedContractSyncCursor<T>
+where
+    T: Send + Debug + 'static,
+{
+    async fn next_action(&mut self) -> Result<(CursorAction, Duration)> {
+        let eta = self.sync_eta()?;
 
         let rate_limit = self.get_rate_limit().await?;
         if let Some(rate_limit) = rate_limit {
             return Ok((CursorAction::Sleep(rate_limit), eta));
         }
-        let (max_sequence, tip) = self.indexer.latest_sequence_count_and_tip().await?;
-        self.tip = tip;
-        self.max_sequence = max_sequence;
-        if let Some(range) = self.sync_state.get_next_range(max_sequence, tip).await? {
+
+        if let Some(range) = self.get_next_range().await? {
             return Ok((CursorAction::Query(range), eta));
+        } else {
+            // TODO: Define the sleep time from interval flag
+            return Ok((CursorAction::Sleep(Duration::from_secs(5)), eta));
         }
-
-        // TODO: Define the sleep time from interval flag
-        Ok((CursorAction::Sleep(Duration::from_secs(5)), eta))
     }
 
     fn latest_queried_block(&self) -> u32 {
         self.sync_state.next_block.saturating_sub(1)
     }
 
-    async fn update(&mut self, _: Vec<(T, LogMeta)>, _range: RangeInclusive<u32>) -> Result<()> {
+    async fn update(&mut self, _: Vec<(T, LogMeta)>, range: RangeInclusive<u32>) -> Result<()> {
```

Comment:
what should we do for Cosmos IGP indexing? Atm `fetch_logs` doesn't propagate any errors that we may get along the way, so we end up calling update here and move onto the next range even if we skipped some blocks https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/dan/watermark-refactoring/rust/chains/hyperlane-cosmos/src/interchain_gas.rs#L231

Open to this being addressed in a different PR but we should try to have a consistent interface between chain log fetching impls when it comes to errors

Reply:
pushed a fix for this, lmk if it lgty